File: backend/api/routes/literature.py
"""Literature search and ingest API routes."""

import logging

from fastapi import APIRouter, Query

from backend.api.errors import BadRequestError
from backend.pipeline.literature.models import Paper

logger = logging.getLogger(__name__)

router = APIRouter()


class IngestRequest(Paper):
    """Ingest request inherits all Paper fields — title is required for confirmation (HB-01)."""

    pass


# Lazy-initialised singleton to avoid constructing sources at import time
_service = None


def _get_service():
    global _service
    if _service is None:
        from backend.pipeline.literature.search_service import SearchService

        _service = SearchService()
    return _service


@router.get(
    "/search",
    summary="Search academic literature",
    description="Search across multiple academic sources (Semantic Scholar, arXiv, OpenAlex) with deduplication.",
)
async def search_literature(
    q: str = Query(..., description="Search query string"),
    max_results: int = Query(default=10, ge=1, le=100, description="Maximum number of results"),
):
    """Search academic literature across multiple sources.

    Args:
        q: Search query string (required).
        max_results: Maximum number of results to return.

    Returns:
        {"papers": Paper[]}

    Example response:
        {"papers": [{"id": "p1", "source": "arxiv", "title": "Attention Is All You Need", "abstract": "...", "year": 2017}]}
    """
    service = _get_service()
    papers = await service.search_all(query=q, limit_per_source=max_results)
    # Cap to max_results after deduplication
    papers = papers[:max_results]
    return {"papers": papers}


@router.get(
    "/ingested",
    summary="List ingested paper IDs",
    description=(
        "Returns the set of paper IDs that have been ingested into the "
        "knowledge base. Used by the literature UI to derive an "
        "authoritative persisted-ingestion indicator from backend state, "
        "rather than from ephemeral client state."
    ),
)
async def list_ingested_papers():
    """Return the set of paper IDs persisted in the knowledge base.

    The vector store records each ingested paper chunk with metadata
    ``{"source": "academic_paper", "paper_id": <id>}`` (see
    ``_do_ingest`` in this module). We query the collection metadata
    and return the unique paper IDs.

    Returns:
        {"ids": ["p1", "p2", ...]}

    If the vector store is unavailable or the embedding provider is
    offline, we return an empty list rather than failing — the UI
    gracefully treats this as "no persisted ingestion state known".
    """
    try:
        from backend.config import get_settings
        from backend.pipeline.knowledge.embedding_service import EmbeddingService
        from backend.pipeline.knowledge.vector_store import VectorStore
        from backend.providers.provider_factory import create_provider

        settings = get_settings()
        provider = create_provider()
        embedding = EmbeddingService(provider)
        store = VectorStore(settings.chroma_persist_dir, embedding)

        all_meta = store._collection.get(include=["metadatas"])
        ids: list[str] = []
        seen: set[str] = set()
        for m in (all_meta.get("metadatas") or []):
            if not m:
                continue
            # Two ingestion paths write to the vector store:
            #   - add_papers (vector_store.py): metadata type absent, paper_id present
            #   - _do_ingest (this file):       metadata type="academic_paper", paper_id present
            # We accept either as evidence of ingestion.
            paper_id = m.get("paper_id")
            if paper_id and paper_id not in seen:
                seen.add(paper_id)
                ids.append(paper_id)
        return {"ids": ids}
    except Exception as e:
        logger.info("ingested-papers lookup unavailable: %s", e)
        return {"ids": []}


async def _do_ingest(paper: Paper) -> dict:
    """Persist a paper into the vector store with paper_id metadata.

    F1.5d: This is the truthful ingest path. The paper is written via
    VectorStore.add_papers, which embeds the chunk text and upserts into
    the chromadb collection with metadata {"paper_id": paper.id, ...}.
    GET /literature/ingested reads that exact metadata, so the write
    and read paths share one source of truth.

    Failures MUST surface as errors — never return a fake success when
    nothing was persisted. The prior ImportError-fallback was a
    correctness defect (claimed "ingested" without writing anything).
    """
    from backend.api.errors import ServiceUnavailableError
    from backend.config import get_settings
    from backend.pipeline.ingestion.chunker import DocumentChunk
    from backend.pipeline.knowledge.embedding_service import EmbeddingService
    from backend.pipeline.knowledge.vector_store import VectorStore
    from backend.providers.provider_factory import create_provider

    # 1. Build chunk text from paper title + abstract + authors.
    #    One chunk is sufficient for title+abstract papers (matches the
    #    IngestionStage idiom in backend/pipeline/stages.py:511-524).
    parts = [paper.title]
    if paper.abstract:
        parts.append(paper.abstract)
    if paper.authors:
        parts.append(", ".join(a.name for a in paper.authors))
    text = "\n\n".join(p for p in parts if p)

    chunks = [
        DocumentChunk(
            text=text,
            paper_id=paper.id,
            section="abstract",
            chunk_index=0,
        )
    ]

    # 2. Construct the VectorStore exactly like list_ingested_papers and
    #    knowledge.ingest_pdf so the read path observes the same collection.
    try:
        settings = get_settings()
        provider = create_provider()
        embedding = EmbeddingService(provider)
        store = VectorStore(settings.chroma_persist_dir, embedding)
    except Exception as e:
        # Provider/key/chroma not configured — raise, do NOT fake success.
        logger.error("Vector store construction failed for paper %s: %s", paper.id, e)
        raise ServiceUnavailableError(
            detail=f"Knowledge base unavailable: {e}",
            hint="Set EROCK_OPENAI_API_KEY (or switch EROCK_EMBEDDING_PROVIDER to ollama) and ensure chromadb is installed.",
        ) from e

    # 3. Persist. add_papers embeds, dedupes by "{paper.id}_chunk_{i}",
    #    and writes metadata {paper_id, paper_title, source, section,
    #    year, keywords}. Returns the number of chunks upserted.
    try:
        stored = await store.add_papers([paper], [chunks])
    except Exception as e:
        logger.error("Ingestion failed for paper %s: %s", paper.id, e)
        raise BadRequestError(
            detail=f"Ingestion failed: {e}",
            hint="Check that the embedding provider is reachable and the paper has non-empty content.",
        ) from e

    if stored == 0:
        # add_papers silently drops zero-vector chunks (provider offline).
        # Surface this as failure so clients know nothing was persisted
        # and the GET /literature/ingested read path will not report the id.
        raise BadRequestError(
            detail="Ingestion wrote 0 chunks — embedding provider may be offline",
            hint="Switch EROCK_EMBEDDING_PROVIDER to a working provider (openai/gemini/ollama).",
        )

    return {"status": "ingested", "id": paper.id, "chunks": stored}


@router.post(
    "/ingest",
    summary="Ingest paper into knowledge base",
    description="Store a paper in the knowledge base for future reference. Requires user confirmation (title must be present).",
)
async def ingest_paper(paper: IngestRequest):
    """Ingest a paper into the knowledge base.

    HB-01: Ingestion requires user confirmation — the paper must have a title
    to be considered a valid ingestion request.

    Args:
        paper: Paper data to ingest.

    Returns:
        {"status": "ingested", "id": "..."}

    Example request:
        {"id": "p1", "source": "arxiv", "title": "Attention Is All You Need", "abstract": "...", "year": 2017}

    Example response:
        {"status": "ingested", "id": "p1"}
    """
    # HB-01: Validate that the paper has a title (user confirmation requirement)
    if not paper.title or not paper.title.strip():
        raise BadRequestError(
            detail="Paper title is required for ingestion confirmation",
            hint="Ensure the paper has a title before ingesting. This acts as user confirmation.",
        )

    return await _do_ingest(paper)

File: frontend/src/api/literature.ts
import {
  callContract,
  decodeArray,
  decodeNumber,
  decodeObject,
  decodeString,
  type JsonContract,
} from "./contracts/common";

// ── Types ──

export interface Author {
  name: string;
  id?: string | null;
  affiliations?: string[];
}

export interface Paper {
  id: string;
  source: string;
  title: string;
  abstract: string | null;
  authors: Author[];
  year: number | null;
  venue: string | null;
  citation_count: number | null;
  url: string | null;
  doi: string | null;
  arxiv_id: string | null;
  keywords: string[];
}

export interface SearchResponse {
  papers: Paper[];
}

/** Response shape of GET /literature/ingested. */
export interface IngestedPapersResponse {
  ids: string[];
}

// F1.1 M4: renamed from IngestResponse to LiteratureIngestResponse to
// disambiguate from the knowledge-ingest IngestResponse in api/types.ts
// (which has a different shape: {status, filename, chunks}). The two
// endpoints (/literature/ingest vs /knowledge/ingest) genuinely return
// different fields — the name collision was a maintenance trap.
export interface LiteratureIngestResponse {
  status: string;
  id: string;
}

// ── API Calls ──

// F1.7a: contract-validated literature search. The paper-card component
// reads paper.id/title/abstract/authors[].name/source/year/citation_count/
// doi/url (see field-usage survey); these are the material fields. Nullable
// fields (abstract, year, venue, citation_count, url, doi, arxiv_id) are
// validated when present and preserved as null otherwise. authors[].name is
// the only author field consumed by the card, but id/affiliations are
// declared on the type and validated when present.
const authorDecoder = decodeObject<Author>({
  required: { name: decodeString },
  optional: { id: decodeString },
});

const paperDecoder = decodeObject<Paper>({
  required: {
    id: decodeString,
    source: decodeString,
    title: decodeString,
    authors: decodeArray(authorDecoder),
    keywords: decodeArray(decodeString),
  },
  optional: {
    abstract: decodeString,
    year: decodeNumber,
    venue: decodeString,
    citation_count: decodeNumber,
    url: decodeString,
    doi: decodeString,
    arxiv_id: decodeString,
  },
});

const searchLiteratureContract: JsonContract<SearchResponse> = {
  id: "literature.searchLiterature",
  method: "GET",
  pathPattern: "/literature/search",
  responseKind: "json",
  decoder: decodeObject<SearchResponse>({
    required: { papers: decodeArray(paperDecoder) },
  }),
};

/** Search academic literature across multiple sources. */
export function searchLiterature(
  query: string,
  maxResults: number = 10,
): Promise<SearchResponse> {
  // F1.7a: migrated from apiFetchUnchecked to callContract with runtime decoder.
  return callContract(searchLiteratureContract, {
    query: { q: query, max_results: maxResults },
  });
}

// F1.4.1: contract-validated ingest — no longer uses apiFetchUnchecked
const ingestPaperContract: JsonContract<LiteratureIngestResponse> = {
  id: "literature.ingestPaper",
  method: "POST",
  pathPattern: "/literature/ingest",
  responseKind: "json",
  decoder: decodeObject<LiteratureIngestResponse>({
    required: { status: decodeString, id: decodeString },
  }),
};

/** Ingest a paper into the knowledge base. */
export function ingestPaper(paper: Paper): Promise<LiteratureIngestResponse> {
  return callContract(ingestPaperContract, { body: paper });
}

// F1.5c: contract-validated ingested-papers read — authoritative source of
// persisted ingestion state. The literature UI derives its "Ingested" badge
// from this response rather than from ephemeral client state, so the badge
// survives reload/remount and always reflects backend truth.
const ingestedPapersContract: JsonContract<IngestedPapersResponse> = {
  id: "literature.listIngested",
  method: "GET",
  pathPattern: "/literature/ingested",
  responseKind: "json",
  decoder: decodeObject<IngestedPapersResponse>({
    required: { ids: decodeArray(decodeString) },
  }),
};

/** Fetch the authoritative set of ingested paper IDs from the backend. */
export function listIngestedPapers(): Promise<IngestedPapersResponse> {
  return callContract(ingestedPapersContract);
}
